src/analysis/processing/qgsalgorithmjoinbylocation.cpp

```diff
@@ -0,0 +1,368 @@
+{
+  return QObject::tr( "Join attributes by location" );
+}
+
+QStringList QgsJoinByLocationAlgorithm::tags() const
+{
+  return QObject::tr( "join,intersects,intersecting,touching,within,contains,overlaps,relation,spatial" ).split( ',' );
+}
+
+QString QgsJoinByLocationAlgorithm::group() const
+{
+  return QObject::tr( "Vector general" );
+}
+
+QString QgsJoinByLocationAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectorgeneral" );
+}
+
+QString QgsJoinByLocationAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm takes an input vector layer and creates a new vector layer "
+                      "that is an extended version of the input one, with additional attributes in its attribute table.\n\n"
+                      "The additional attributes and their values are taken from a second vector layer. "
+                      "A spatial criteria is applied to select the values from the second layer that are added "
+                      "to each feature from the first layer in the resulting one." );
+}
+
+QString QgsJoinByLocationAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Join attributes from one vector layer to another by location." );
+}
+
+QgsJoinByLocationAlgorithm *QgsJoinByLocationAlgorithm::createInstance() const
+{
+  return new QgsJoinByLocationAlgorithm();
+}
+
+bool QgsJoinByLocationAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  mBaseSource.reset( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !mBaseSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  mJoinSource.reset( parameterAsSource( parameters, QStringLiteral( "JOIN" ), context ) );
+  if ( !mJoinSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "JOIN" ) ) );
+
+
+  mJoinMethod = parameterAsEnum( parameters, QStringLiteral( "METHOD" ), context );
+
+  mPredicates = parameterAsEnums( parameters, QStringLiteral( "PREDICATE" ), context );
+
+  QString prefix = parameterAsString( parameters, QStringLiteral( "PREFIX" ), context );
+
+  if ( mFieldNames.empty() )
+  {
+    outFields2 = mJoinSource->fields();
+    fields2Indices.reserve( outFields2.count() );
+    for ( int i = 0; i < outFields2.count(); ++i )
+    {
+      fields2Indices << i;
+    }
+  }
+  else
+  {
+    fields2Indices.reserve( mFieldNames.count() );
+    for ( const QString &field : mFieldNames )
+    {
+      int index = mJoinSource->fields().lookupField( field );
+      if ( index >= 0 )
+      {
+        fields2Indices << index;
+        outFields2.append( mJoinSource->fields().at( index ) );
+      }
+    }
+  }
+
+  if ( !prefix.isEmpty() )
+  {
+    for ( int i = 0; i < outFields2.count(); ++i )
+    {
+      outFields2[ i ].setName( prefix + outFields2[ i ].name() );
+    }
+  }
+
+  if ( feedback->isCanceled() )
```

Comment:
You can drop this -- the prepareAlgorithm step is all cheap code, so we don't need to support cancelation at this stage